### PR TITLE
ENH: spatial.transform: improve performance and fix gradients of `Rotation.from_matrix`

### PR DESCRIPTION
**Pull request by amacati**
### Issue
This PR is not long, but the changes are a little subtle, so I think it's worth motivating where it comes from. I noticed something strange while working with `Rotation` in JAX using the array API. Gradients that flow through `from_matrix` are NaN, even though the problem is perfectly differentiable. The only candidate for trouble in that function is the SVD, which should be differentiable.

Ironically, the SVD **is** differentiable, except if the matrix has duplicate singular values. Rotations, being volume-preserving, have three singular values of 1. So **valid** rotation matrices as input produce the NaNs here, even though we discard the result later. I first assumed that the masking protects us against propagating NaNs. But even after fixing the orthogonalization threshold for `float32` inputs (even random, valid rotation matrices do not pass the test up to 1e-6 tolerance) I still saw NaNs.

However, it turns out that selecting via `xp.where` can in fact propagate NaN gradients, even if the result is discarded (see [JAX documentation](https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where)). The fix is a second select that already filters the input arguments and breaks the auto-diff link between the input and the function that is producing NaNs.

At this point I was a bit worried, since the function as it is now is perfectly valid, and we probably don't want to introduce complexities in scipy that fix behaviors from other frameworks. However, it turns out that our choice of replacement can also improve the performance of `from_matrix`. We discard the results of already orthogonal matrices anyways, so we are free to replace the matrix entries with anything. The easiest choice is a unit matrix. The SVD is an iterative algorithm, even if it runs on lazy frameworks with otherwise static computation graphs like JAX. And since the unit matrix, being perfectly diagonal, is a trivial case for the SVD to solve, this allows the algorithm to immediately early-stop in those cases. So here we are with another performance optimization PR that happens to fix gradients as well :D

### Reference issue
Vaguely related to #25691

### What does this implement/fix?
Update the orthogonalization threshold in `from_matrix` to be dtype-adaptive. Set matrices that are confirmed to be orthogonal and normal to the unit matrix on lazy frameworks to accelerate the branch that is taken, but discarded, and to fix the resulting gradients on autodiff frameworks.

### Benchmarks
This PR only concerns JAX, because it is the only lazy framework I test in my benchmark suite. The others do not take this branch.
from_matrix_svd

### AI Generation Disclosure
Claude for benchmarking and initial scan of `from_matrix`

### CODE REVIEW

**Review by lucascolley — APPROVED** [review]
thanks for the detailed explanation! All LGTM

File: scipy/spatial/transform/_rotation_xp.py [hunk 1]
```diff
@@ -65,15 +65,30 @@ def from_matrix(matrix: Array, assume_valid: bool = False) -> Array:
 
         gramians = matrix @ matrix.mT
         eye = xp.eye(3, dtype=matrix.dtype, device=device)
+        # float32 reduced precision compared to float64 requires a higher threshold
+        atol = 1e-12 if matrix.dtype == xp.float64 else 1e-6
```
Comment (scottshambaugh) [inline scipy/spatial/transform/_rotation_xp.py:69]: Checking my notes (https://github.com/scipy/scipy/pull/24056), we still do correct for a bit of small-angle orthogonalization error in the `_from_matrix_orthogonal` calcs, so this is fine.

**Review by scottshambaugh — APPROVED** [review]
Looks like getting into the weeds is paying some dividends, thanks for digging into this!
